src/libaktualizr/storage/sqlstorage.cc

```diff
@@ -1756,11 +1817,25 @@ class SQLTargetWHandle : public StorageTargetWHandle {
     if (stream_) {
       stream_.close();
     }
-    if (sqlite3_changes(db_.get()) > 0) {
-      auto statement =
-          db_.prepareStatement<std::string>("DELETE FROM target_images WHERE targetname=?;", target_.filename());
-      if (statement.step() != SQLITE_DONE) {
-        LOG_ERROR << "could not delete " << target_.filename() << " from sql storage";
+
+    if (storage_ != nullptr) {
+      SQLite3Guard db = storage_->dbConnection();
+      ;
+      if (sqlite3_changes(db.get()) > 0) {
```

Comment:
It looks like `sqlite3_changes` will always return 0 in this case, as the database connection has just been created: https://www.sqlite.org/c3ref/changes.html.

We have to find another way to track that here, maybe another attribute would be enough.

Reply:
Ah, yes, it's a good point. And I found the "insert/replace" is in one of the constructors of WHandle class, so in this case we could run "delete" anyway in abort() function. And in another constructor which is basically for test code, there is nothing to do with the table in database, so we don't need to run "delete" in this case. Please correct me if I made something wrong.